test/tools/yulInterpreter/EVMInstructionInterpreter.cpp

```diff
@@ -0,0 +1,489 @@
+	else
+	{
+		size_t off = size_t(_offset);
+		u256 val;
+		for (size_t i = 0; i < 32; ++i)
+		{
+			val <<= 8;
+			if (off + i < _data.size())
+				val += _data[off + i];
+		}
+		return val;
+	}
+}
+
+void copyZeroExtended(
+	bytes& _target, bytes const& _source,
+	size_t _targetOffset, size_t _sourceOffset, size_t _size
+)
+{
+	yulAssert(_targetOffset + _size < _target.size(), "");
+	for (size_t i = 0; i < _size; ++i)
+		_target[_targetOffset + i] = _sourceOffset + i < _source.size() ? _source[_sourceOffset + i] : 0;
+}
+
+}
+
+using u512 = boost::multiprecision::number<boost::multiprecision::cpp_int_backend<512, 256, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>;
+
+u256 EVMInstructionInterpreter::eval(
+	solidity::Instruction _instruction,
+	vector<u256> const& _arguments
+)
+{
+	using dev::solidity::Instruction;
+
+	auto info = solidity::instructionInfo(_instruction);
+	yulAssert(size_t(info.args) == _arguments.size(), "");
+
+	auto const& arg = _arguments;
+	switch (_instruction)
+	{
+	case Instruction::STOP:
+		throw InterpreterTerminated();
+	// --------------- arithmetic ---------------
+	case Instruction::ADD:
+		return arg[0] + arg[1];
+	case Instruction::MUL:
+		return arg[0] * arg[1];
+	case Instruction::SUB:
+		return arg[0] - arg[1];
+	case Instruction::DIV:
+		return arg[1] == 0 ? 0 : arg[0] / arg[1];
+	case Instruction::SDIV:
+		return arg[1] == 0 ? 0 : s2u(u2s(arg[0]) / u2s(arg[1]));
+	case Instruction::MOD:
+		return arg[1] == 0 ? 0 : arg[0] % arg[1];
+	case Instruction::SMOD:
+		return arg[1] == 0 ? 0 : s2u(u2s(arg[0]) % u2s(arg[1]));
+	case Instruction::EXP:
+		return exp256(arg[0], arg[1]);
+	case Instruction::NOT:
+		return ~arg[0];
+	case Instruction::LT:
+		return arg[0] < arg[1] ? 1 : 0;
+	case Instruction::GT:
+		return arg[0] > arg[1] ? 1 : 0;
+	case Instruction::SLT:
+		return u2s(arg[0]) < u2s(arg[1]) ? 1 : 0;
+	case Instruction::SGT:
+		return u2s(arg[0]) < u2s(arg[1]) ? 1 : 0;
+	case Instruction::EQ:
+		return arg[0] == arg[1] ? 1 : 0;
+	case Instruction::ISZERO:
+		return arg[0] == 0 ? 1 : 0;
+	case Instruction::AND:
+		return arg[0] & arg[1];
+	case Instruction::OR:
+		return arg[0] | arg[1];
+	case Instruction::XOR:
+		return arg[0] ^ arg[1];
+	case Instruction::BYTE:
+		return arg[0] >= 32 ? 0 : (arg[1] >> unsigned(8 * (31 - (arg[0] & 0xff)))) & 0xff;
+	case Instruction::SHL:
+		return arg[0] > 255 ? 0 : (arg[1] << unsigned(arg[0] & 0xff));
+	case Instruction::SHR:
+		return arg[0] > 255 ? 0 : (arg[1] >> unsigned(arg[0] & 0xff));
+	case Instruction::SAR:
+	{
+		static u256 const hibit = u256(1) << 255;
+		if (arg[0] >= 256)
+			return arg[1] & hibit ? u256(-1) : 0;
+		else
+		{
+			unsigned amount = unsigned(arg[0] & 0xff);
+			u256 v = arg[1] >> amount;
+			if (arg[1] & hibit)
+				v |= u256(-1) << (256 - amount);
+			return v;
+		}
+	}
+	case Instruction::ADDMOD:
+		return arg[2] == 0 ? 0 : u256((u512(arg[0]) + u512(arg[1])) % arg[2]);
```

Comment:
We might be able to avoid the upcast to ``u512`` here... but I haven't thought it through, because I'm not sure it's worth it :-)...